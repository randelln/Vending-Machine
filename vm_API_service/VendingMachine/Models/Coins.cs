using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace VendingMachine.Models
{
    public class Coins
    {
        public int value { get; set; }
       public string name { get; set; }
        public int amount { get; set; }
    }
}
