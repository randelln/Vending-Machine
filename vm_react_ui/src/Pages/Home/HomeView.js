import React from 'react';
import { Button, Col, Container, Row } from 'reactstrap';
import DataTable from '../../components/DataTable/DataTableView';
import './HomeStyles.css';

const HomeView = ({ functions, state }) => {
  //Destructuring
 // const object = {name:'Randell' , value:3}
  //const name = object.name
  //const {name} = object
  // 
    const { loadData, fetchCoins, fetchProducts, addCoins } = functions;
    const {
        products,
        coins,
        loadingProducts,
        loadingCoins,
        reLoadingProducts,
        reLoadingCoins,
        loadingBuy,
        setLoadingBuy,
        runningTotal,
        setRunningTotal
    } = state;

    return !loadingBuy && !reLoadingProducts && !reLoadingCoins ? (
        <Container className="main-container" style={{ paddingTop: '10px' }}>
            <Row>
                <Col>
                    <h3>Vending Machine</h3>
                </Col>
                <Col>
                    <h3>Balance: R{runningTotal}</h3>
                </Col>
            </Row>
            <Row>
                <Col>
                    <DataTable
                        products={products}
                        runningTotal={runningTotal}
                        setLoadingBuy={setLoadingBuy}
                        setRunningTotal={setRunningTotal}
                        loadingProducts={loadingProducts}
                        fetchProducts={fetchProducts}
                        fetchCoins={fetchCoins}
                    />
                </Col>
            </Row>
            <tbody>
            {!loadingCoins ? (
                coins.map(coin => (
                    <tc className="button-padding" key={coin.name}>
                        <Button disabled={coin.amount === 0} color="success" onClick={() => addCoins(coin.value)}>
                            Insert {coin.name} ({coin.amount})
                            <br />
                        </Button>
                    </tc>
                ))
            ) : (
                <div className="loader-coins">
                    <img src="https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif" />
                </div>
            )}
            </tbody>

            <br />
            <tbody>
            <tr>
                <Button id="ReloadProductsButton" className="btn-padding-right" color="success" onClick={fetchProducts}>
                    Refresh products
                </Button>

                <Button id="ReloadCoinsButton" className="btn-padding-right" color="success" onClick={() => {
                    fetchCoins();
                    setRunningTotal(0)
                }}>
                    Reload coins
                </Button>

                <Button id="ReloadAll" className="btn-padding-right" color="success" onClick={loadData}>
                    Refresh All
                </Button>

                <Button id="CancelButton" className="btn-padding-right" color="danger" onClick={() => { 
                    setRunningTotal(0); 
                    fetchCoins()}}>
                    Cancel
                </Button>
            </tr>
            </tbody>
        </Container>
    ) : (
        <div className="loader-container" id="HomeMainLoader">
            <img src="https://upload.wikimedia.org/wikipedia/commons/b/b1/Loading_icon.gif" />
        </div>
    );
};

export default HomeView;
