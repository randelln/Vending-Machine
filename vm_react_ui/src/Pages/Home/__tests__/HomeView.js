import React from 'react';
import { shallow } from 'enzyme';
import HomeView from "../HomeView";
import Coins from './__data__/Coins.json';
import Products from './__data__/Products.json';

describe('HomeView', () => {

    const props = {
        functions: {
            loadData: jest.fn(),
            fetchCoins: jest.fn(),
            fetchProducts: jest.fn(),
            addCoins: jest.fn(),
        },
        state: {
            products: Products,
            coins: Coins,
            loadingProducts: false,
            loadingCoins: false,
            reLoadingProducts: false,
            reLoadingCoins: false,
            loadingBuy: false,
            setLoadingBuy: jest.fn(),
            runningTotal: 10,
            setRunningTotal: jest.fn(),
        }
    }

    const view = shallow(<HomeView {...props} />).dive();

    it('should render the HomeView without issues', () => {
        expect(view).toMatchSnapshot();
    });

    it('should have no loader when loading is false', () => {
        const loader = view.findWhere(el => el.props().id === 'HomeMainLoader');
        expect(loader).toHaveLength(0);
    });

    it('should have the loader when loading is true', () => {
        const negativeProps = props;
        negativeProps.state.loadingBuy = true;
        negativeProps.state.reLoadingProducts = true;
        negativeProps.state.reLoadingCoins = true;
        const view = shallow(<HomeView {...negativeProps} />);
        const loader = view.findWhere(el => el.props().id === 'HomeMainLoader');
        expect(loader).toHaveLength(1);
    });

    it('should call fetchProducts() when ReloadProductsButton is pressed', () => {
        view.findWhere(el => el.props().id === 'ReloadProductsButton').simulate('click');
        expect(props.functions.fetchProducts).toHaveBeenCalled();
    });

    it('should call fetchCoins() when ReloadCoinsButton is pressed', () => {
        view.findWhere(el => el.props().id === 'ReloadCoinsButton').simulate('click');
        expect(props.functions.fetchCoins).toHaveBeenCalled();
    });

    it('should call loadData() when ReloadAll is pressed', () => {
        view.findWhere(el => el.props().id === 'ReloadAll').simulate('click');
        expect(props.functions.loadData).toHaveBeenCalled();
    });

    it('should reset the running total when CancelButton is pressed', () => {
        view.findWhere(el => el.props().id === 'CancelButton').simulate('click');
        expect(props.state.setRunningTotal).toHaveBeenCalledWith(0);
    });
});
