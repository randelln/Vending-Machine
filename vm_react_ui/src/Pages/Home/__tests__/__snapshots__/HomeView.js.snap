// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView should render the HomeView without issues 1`] = `
<div
  className="main-container container"
  style={
    Object {
      "paddingTop": "10px",
    }
  }
>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h3>
        Vending Machine
      </h3>
    </Col>
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h3>
        Balance: R
        10
      </h3>
    </Col>
  </Row>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <DataTable
        fetchCoins={[MockFunction]}
        fetchProducts={[MockFunction]}
        loadingProducts={false}
        products={
          Array [
            Object {
              "id": 1,
              "name": "Coke",
              "price": 10,
              "quantity": 10,
            },
            Object {
              "id": 2,
              "name": "Fanta",
              "price": 10,
              "quantity": 2,
            },
            Object {
              "id": 3,
              "name": "Sprite",
              "price": 10,
              "quantity": 3,
            },
            Object {
              "id": 4,
              "name": "Lemon Twist",
              "price": 10,
              "quantity": 9,
            },
            Object {
              "id": 5,
              "name": "Tab",
              "price": 10,
              "quantity": 5,
            },
            Object {
              "id": 6,
              "name": "Pepsi",
              "price": 10,
              "quantity": 8,
            },
            Object {
              "id": 7,
              "name": "KitKat",
              "price": 8,
              "quantity": 7,
            },
            Object {
              "id": 8,
              "name": "Lays",
              "price": 5,
              "quantity": 12,
            },
            Object {
              "id": 9,
              "name": "Simba",
              "price": 5,
              "quantity": 9,
            },
            Object {
              "id": 10,
              "name": "Energade",
              "price": 10,
              "quantity": 11,
            },
          ]
        }
        runningTotal={10}
        setLoadingBuy={[MockFunction]}
        setRunningTotal={[MockFunction]}
      />
    </Col>
  </Row>
  <tbody>
    <tc
      className="button-padding"
      key="R1"
    >
      <Button
        color="success"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Insert 
        R1
         (
        )
        <br />
      </Button>
    </tc>
    <tc
      className="button-padding"
      key="R2"
    >
      <Button
        color="success"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Insert 
        R2
         (
        )
        <br />
      </Button>
    </tc>
    <tc
      className="button-padding"
      key="R5"
    >
      <Button
        color="success"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Insert 
        R5
         (
        )
        <br />
      </Button>
    </tc>
    <tc
      className="button-padding"
      key="R10"
    >
      <Button
        color="success"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Insert 
        R10
         (
        )
        <br />
      </Button>
    </tc>
  </tbody>
  <br />
  <tbody>
    <tr>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadProductsButton"
        onClick={[MockFunction]}
        tag="button"
      >
        Refresh products
      </Button>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadCoinsButton"
        onClick={[Function]}
        tag="button"
      >
        Reload coins
      </Button>
      <Button
        className="btn-padding-right"
        color="success"
        id="ReloadAll"
        onClick={[MockFunction]}
        tag="button"
      >
        Refresh All
      </Button>
      <Button
        className="btn-padding-right"
        color="danger"
        id="CancelButton"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
    </tr>
  </tbody>
</div>
`;
